### ISSUE TITLE
ERROR:  invalid attribute number <number>

### ISSUE THREAD

**Issue by AndyMender** [timescale/timescaledb#526]
I started using TimescaleDB some months ago and recently I noticed a lot of failed `INSERT INTO...` queries with an enigmatic PostgreSQL error: `ERROR: invalid attribute number <number>` where `number` can differ depending on the table and database.

Initially, it was happening for only 1 table type in all databases, now it's happening for all tables in some databases.

Today I upgraded to TimescaleDB 0.9.2 and the problem still persists (PostgreSQL version 9.6.x). I tried running `VACUUM FULL` and` REINDEX`, but that didn't help either.

Most reported cases point to an error in the syntax of the query, in addition to the error shown above. This is not the case for me and I have carefully inspected the query without observing any issues with it.

**EDIT:** I tried validating my queries again and I get the above error even when inserting a single row.

**Comment by AndyMender** [comment]
@RobAtticus actually, I managed to address my issue by performing a full `pg_dump` and `pg_restore` for all of our databases. Most likely the table indices got corrupted when I was rebuilding constraints and adding/dropping some columns around a month ago. I'm still a bit new to TimescaleDB.

As for the other relevant information:
OS: OpenSUSE 42.3 LEAP
Architecture: 64-bit (KVM virtual machine on Ubuntu 16.04 LTS)
PostgreSQL: 9.6.8

Since the problem was actually solved, you can close this. However, it's probably worth noting the quirk someplace :).

**Comment by AndyMender** [comment]
@davidkohn88 the data I typically deal with belongs to our clients, so I'm not at liberty to share it, naturally. I'm not sure how to reproduce this behavior myself. The column add/drop was merely a hunch and I can of course be wrong. I haven't seen anyone else experiencing this behavior either.

**Comment by mfreed** [comment]
Hi @AndyMender have you ever seen this behavior again since?

**Comment by AndyMender** [comment]
@mfreed actually, yes. I linked this issue to a PR that might be able to address my problem along with those associated with similar `ALTER TABLE hypertable DROP COLUMN ...;` query. Thank you kindly for the assistance! :)

**Comment by Ngalstyan4** [comment]
@AndyMender thanks for the feedback! I was able to reproduce this and will take a look later today.

**Comment by AndyMender** [comment]
Per an e-mail request from @Ngalstyan4 here is the table schema for one of the affected tables:
```
                                 Table "public.basics"
  Column  |           Type           | Modifiers | Storage | Stats target | Description 
----------+--------------------------+-----------+---------+--------------+-------------
 time     | timestamp with time zone | not null  | plain   |              | 
 EURO     | double precision         |           | plain   |              | 
 E_Z_CAB4 | double precision         |           | plain   |              | 
 E_Z_EVU  | double precision         |           | plain   |              | 
 E_Z_PV3  | double precision         |           | plain   |              | 
 E_Z_PV4  | double precision         |           | plain   |              | 
 E_Z_SOLL | double precision         |           | plain   |              | 
 G_M0     | double precision         |           | plain   |              | 
Indexes:
    "basics_time_key" UNIQUE CONSTRAINT, btree ("time")
Child tables: _timescaledb_internal._hyper_10_100_chunk,
              _timescaledb_internal._hyper_10_104_chunk,
              _timescaledb_internal._hyper_10_108_chunk,
              _timescaledb_internal._hyper_10_112_chunk,
              _timescaledb_internal._hyper_10_116_chunk,
              _timescaledb_internal._hyper_10_120_chunk,
              _timescaledb_internal._hyper_10_124_chunk,
              _timescaledb_internal._hyper_10_128_chunk,
              _timescaledb_internal._hyper_10_132_chunk,
              _timescaledb_internal._hyper_10_136_chunk,
              _timescaledb_internal._hyper_10_140_chunk,
              _timescaledb_internal._hyper_10_144_chunk,
              _timescaledb_internal._hyper_10_148_chunk,
              _timescaledb_internal._hyper_10_152_chunk,
              _timescaledb_internal._hyper_10_156_chunk,
              _timescaledb_internal._hyper_10_160_chunk,
              _timescaledb_internal._hyper_10_164_chunk,
              _timescaledb_internal._hyper_10_168_chunk,
              _timescaledb_internal._hyper_10_172_chunk,
              _timescaledb_internal._hyper_10_176_chunk,
              _timescaledb_internal._hyper_10_185_chunk,
              _timescaledb_internal._hyper_10_188_chunk,
              _timescaledb_internal._hyper_10_193_chunk,
              _timescaledb_internal._hyper_10_197_chunk,
```
Usually there is a `UNIQUE CONSTRAINT` on the "time" column or a combined `UNIQUE CONSTRAINT` on "time" and "id". The titular error occurs some time after any of the __other__ columns are removed via `ALTER TABLE basics DROP COLUMN` (since "time" and "id" are mandatory). Interestingly, the error doesn't occur always, but rather sporadically.

Mind, this was last tested on PostgreSQL 9.6.9 and TimescaleDB 0.9.2 (I know, not the most recent stable version) on openSUSE 42.3 Leap a while back. Since we're doing a bit of a migration to Ubuntu 18.04 LTS soonish (PostgreSQL 10.x and TimescaleDB 1.x), I'll update this issue once that is done :).

**Comment by Raphyyy** [comment]
I am facing the same trouble here with latest TimescaleDB.
The problem appear with 0.9.2 (the db was created with 0.9.2). After updating to the latest, nothing change.

When I perform this statement : 
`insert into "events" ("id_event_type", "id_sensor") values (14, '998') returning *`

I only have this error : 
`[Err] ERROR:  invalid attribute number 6`

Here is how the table has been created : 

```
CREATE SEQUENCE "events_id_seq"
 INCREMENT 1
 MINVALUE 1
 MAXVALUE 9223372036854775807
 START 74888
 CACHE 1;
SELECT setval('"public"."events_id_seq"', 74888, true);

CREATE TABLE "events" (
"id" int4 DEFAULT nextval('events_id_seq'::regclass) NOT NULL,
"id_sensor" int8 REFERENCES sensors (id) ON DELETE CASCADE ON UPDATE CASCADE NOT NULL,
"id_event_type" int4 REFERENCES event_type (id) ON DELETE CASCADE ON UPDATE CASCADE NOT NULL,
"datetime" timestamp DEFAULT now() NOT NULL,
"comment" text
)
WITH (OIDS=FALSE);
SELECT create_hypertable('events', 'datetime', chunk_time_interval => interval '1 week');
CREATE INDEX ON events (id_sensor, datetime DESC);
CREATE INDEX ON events (id_event_type, datetime DESC);
```

I have been running the table well for months and recently I have updated the field "comment" replacing its type *text* by *jsonb*.
It running well last week then this error appears yersterday.

*The strange part is that if I remove the `returning *` from the statement, everything goes fine.*

This : `insert into "events" ("id_event_type", "id_sensor") values (14, '998')` runs fine and the row is well inserted. 

I have no problem to perform UPDATE or SELECT statement of any kind on the table.
Even this statement : `UPDATE events SET id_sensor = 998 WHERE id = 128734 returning *`
runs well, despite the `returning *`

The error only triggers on : 
`insert into "events" ("id_event_type", "id_sensor") values (14, '998') returning *` 
OR 
`insert into "events" ("comment", "id_event_type", "id_sensor") values ('{"id_last_conf":694}', 14, '998') returning *` 

BUT ! (I just found out)

This statement : 
`insert into "events" ("datetime", "comment", "id_event_type", "id_sensor") values ('2018-01-01', '{"id_last_conf":694}', 14, '998') returning *` also runs well.
Setting the datetime (now() DEFAULT) seems to bypass the problem.

EDIT : 
If I set the datetime to now(), the problem shows up again 
`insert into "events" ("datetime", "comment", "id_event_type", "id_sensor") values (now(), '{"id_last_conf":694}', 14, '998') returning *`
return : 
`[Err] ERROR:  invalid attribute number 6`

But again, this : `insert into "events" ("datetime", "comment", "id_event_type", "id_sensor") values (now(), '{"id_last_conf":694}', 14, '998')` runs well.

**Comment by gayyappan** [comment]
Anyone who runs into this issue , could you give the output for these queries after changing the table_name and schema_name. Thanks.

select count(*) from _timescaledb_catalog.chunk
where hypertable_id = (Select id from _timescaledb_catalog.hypertable
where table_name like 'hourly_sensor_data' and schema_name = 'public') ;


 select column_name, ordinal_position, data_type, column_default, count(*)
from information_schema.columns
where table_name in ( select table_name from _timescaledb_catalog.chunk             
where hypertable_id = (Select id from _timescaledb_catalog.hypertable
where table_name like 'hourly_sensor_data' and schema_name = 'public')
)
group by column_name, ordinal_position, data_type, column_default ;

**Comment by volker48** [comment]
Somehow triggered this bug twice now in one week on the same table. I've been writing to it for a few months without issue.

Here is the output of the requested queries.

```
 count
-------
    32
```

```
-[ RECORD 1 ]----+-------------------------
column_name      | col1
ordinal_position | 3
data_type        | timestamp with time zone
column_default   |
count            | 32
-[ RECORD 2 ]----+-------------------------
column_name      | col2
ordinal_position | 4
data_type        | ARRAY
column_default   |
count            | 32
-[ RECORD 3 ]----+-------------------------
column_name      | col3
ordinal_position | 9
data_type        | boolean
column_default   | false
count            | 32
-[ RECORD 4 ]----+-------------------------
column_name      | col4
ordinal_position | 5
data_type        | ARRAY
column_default   |
count            | 32
-[ RECORD 5 ]----+-------------------------
column_name      | col5
ordinal_position | 10
data_type        | text
column_default   |
count            | 32
-[ RECORD 6 ]----+-------------------------
column_name      | col6
ordinal_position | 11
data_type        | boolean
column_default   |
count            | 32
-[ RECORD 7 ]----+-------------------------
column_name      | col7
ordinal_position | 2
data_type        | character varying
column_default   |
count            | 32
-[ RECORD 8 ]----+-------------------------
column_name      | col8
ordinal_position | 1
data_type        | jsonb
column_default   |
count            | 32
```

Please let me know if you need anymore info from me to help isolate the issue. Thanks in advance.

**Comment by mfreed** [comment]
Hi @volker48 Any updates on your testing?

**Comment by volker48** [comment]
@mfreed we haven't seen it happen since the upgrade so hopefully this is fixed!

**Comment by ByronAP** [comment]
Most of my DB seems messed up
`invalid attribute number -6 for _hyper_24_2188_chunk`
on another table
`ERROR:  invalid attribute number -3`
and again on a different table
`ERROR:  invalid attribute number -1 for _hyper_51_3643_chunk`

Not every table is messed up but a lot are.  
If I try to run the SQL above that @gayyappan shared it fails as well
`ERROR:  value for domain information_schema.cardinal_number violates check constraint "cardinal_number_domain_check"`

the tables are fairly simple, here is a create statement to show structure.
`CREATE TABLE "mytable" (id TIMESTAMP WITH TIME ZONE NOT NULL PRIMARY KEY, o DOUBLE PRECISION, h DOUBLE PRECISION, l DOUBLE PRECISION, c DOUBLE PRECISION, v DOUBLE PRECISION); SELECT create_hypertable('mytable', 'id');`

any help greatly appreciated as I can not do anything with these tables, can't even drop or export them

**Comment by cbrecabarren** [comment]
I'm seeing similar error messages in the process of migrating a database from Postgres 10 to 12. We are using Timescale 1.7.4.

We followed the directions in: https://docs.timescale.com/v1.7/using-timescaledb/backup#backup

On queries we see these types of errors:

```
ERROR:  invalid attribute number 72 for <hypertable>
```

```
ERROR:  could not find inherited attribute "<primary_key_attr>" of relation "_hyper_23_2392_chunk"
```

```
ERROR:  invalid attribute number -1 for _hyper_23_982_chunk
```

On reindex we see these errors:

```
ERROR:  catalog is missing 10 attribute(s) for relid 12446267
```

**Comment by svenklemm** [comment]
@cbrecabarren @ByronAP this is a 2 year old ticket that is closed for almost a year, please open a new ticket with as much information as possible
